codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java

```diff
@@ -0,0 +1,252 @@
+
+        // Add a state for the connection.
+        stateKey = connection.newKey();
+        Http2Stream connectionStream = connection.connectionStream();
+        State connectionState = new State(connectionStream);
+        connectionStream.setProperty(stateKey, connectionState);
+
+        // Register for notification of new streams.
+        connection.addListener(new Http2ConnectionAdapter() {
+            @Override
+            public void onStreamAdded(Http2Stream stream) {
+                State state = new State(stream);
+                stream.setProperty(stateKey, state);
+            }
+
+            @Override
+            public void onStreamClosed(Http2Stream stream) {
+                state(stream).close();
+            }
+        });
+    }
+
+    @Override
+    public void updateStreamableBytes(StreamState streamState) {
+        State state = state(streamState.stream());
+        state.updateStreamableBytes(streamState.streamableBytes(), streamState.hasFrame());
+    }
+
+    @Override
+    public boolean distribute(int maxBytes, Writer writer) throws Http2Exception {
+        checkNotNull(writer, "writer");
+
+        // Allocate the bytes to the streams.
+        long allocated = allocateBytes(maxBytes);
+
+        long unallocatedStreamableBytes = totalStreamableBytes - allocated;
+
+        // Need to write even if maxBytes == 0 in order to handle the case of empty frames.
+        writeVisitor.writeAllocatedBytes(writer);
+
+        return unallocatedStreamableBytes > 0;
+    }
+
+    /**
+     * Allocates the given bytes across the streams.
+     *
+     * @param maxBytes The maximum number of bytes to allocate.
+     * @return the total number of bytes actually allocated.
+     */
+    private long allocateBytes(int maxBytes) {
+        if (queue.isEmpty() || maxBytes <= 0) {
+            return 0L;
+        }
+
+        long allocated = 0;
+        long unallocated = totalStreamableBytes;
+        int allocationChunk = max(minAllocationChunk, (int)(unallocated / queue.size()));
+        while (maxBytes > 0 && !queue.isEmpty()) {
+            //int allocationChunk = max(minAllocationChunk, (int)(unallocated / queue.size()));
+
+            // Remove the head of the queue.
+            State state = queue.remove();
+            state.enqueued = false;
+
+            // Allocate as much data as we can.
+            int chunk = min(allocationChunk, min(maxBytes, state.unallocated()));
+            state.allocate(chunk);
+            maxBytes -= chunk;
+            allocated += chunk;
+            unallocated -= chunk;
+
+            if (state.unallocated() > 0) {
+                // The stream still has unallocated bytes, add it back to the end of the queue.
+                state.enqueue();
+            }
+        }
+
+        return allocated;
+    }
+
+    private State state(Http2Stream stream) {
+        return checkNotNull(stream, "stream").getProperty(stateKey);
+    }
+
+    private class EnqueueStrategy {
```

Comment:
If we only envision there being these 2 strategies can we consider removing this class?  We could consolidate the usage of `state.enqueued`, remove a few types, and remove the ` EnqueueStrategy enqueueStrategy` member for each `State` object:

```java
private final class State {
  private boolean hasBeenEnqueued;
  void enqueue() {
    if (hasBeenEnqueued) {
       queue.addLast(this);
    } else {
       queue.addFirst(this);
       hasBeenEnqueued = true;
    }
    enqueued = true;
  }

  static State dequeue() {
    State s = queue.removeFirst();
    s.enqueued = false;
  }
}
```